pghoard/walreceiver.py

```diff
@@ -0,0 +1,206 @@
+from io import BytesIO
+from pghoard.wal import convert_integer_to_lsn, name_for_tli_log_seg, XLOG_SEG_SIZE
+from psycopg2.extras import PhysicalReplicationConnection, REPLICATION_PHYSICAL  # pylint: disable=no-name-in-module
+from queue import Empty, Queue
+from threading import Thread
+import datetime
+import logging
+import psycopg2
+import select
+
+
+KEEPALIVE_INTERVAL = 10.0
+
+
+class WALReceiver(Thread):
+    def __init__(self, config, connection_string, compression_queue, replication_slot, pg_version_server, site,
+                 last_flushed_lsn=None):
+        super().__init__()
+        self.log = logging.getLogger("WALReceiver")
+        self.running = True
+        self.config = config
+        self.pg_version_server = pg_version_server
+        self.compression_queue = compression_queue
+        self.replication_slot = replication_slot
+        self.completed_wal_segments = set()
+        self.dsn = connection_string
+        self.site = site
+        self.conn = None
+        self.c = None
+        self.buffer = BytesIO()
+        self.latest_wal = None
+        self.latest_wal_start = None
+        self.c = None
+        self.latest_activity = datetime.datetime.utcnow()
+        self.callbacks = {}
+        self.last_flushed_lsn = last_flushed_lsn
+        self.log.info("WALReceiver initialized with replication_slot: %r, last_flushed_lsn: %r",
+                      self.replication_slot, last_flushed_lsn)
+
+    def _get_cursor(self):
+        self.conn = psycopg2.connect(self.dsn, connection_factory=PhysicalReplicationConnection)
+        self.c = self.conn.cursor()
+
+    def create_replication_slot(self):
+        try:
+            self.c.create_replication_slot(self.replication_slot, slot_type=REPLICATION_PHYSICAL)
+        except psycopg2.ProgrammingError:
+            # Slot already exists
```

Comment:
should we check the error code (or message) that this is actually that specific error?